File: csharp/markdown/Markdown.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;

public static class Markdown
{
    private static string Wrap(string text, string tag) => $"<{tag}>{text}</{tag}>";

    private static string ParseText(string markdown, bool isList)
    {
        string Parse(string input, string delimiter, string tag)
        {
            var pattern = $"{delimiter}(.+){delimiter}";
            var replacement = $"<{tag}>$1</{tag}>";
            return Regex.Replace(input, pattern, replacement);
        }

        var parsedText = Parse(Parse(markdown, "__", "strong"), "_", "em");
        return isList ? parsedText : Wrap(parsedText, "p");
    }

    private static (string result, bool isList) ParseHeader(string markdown, bool isList)
    {
        var count = markdown.Length - markdown.TrimStart('#').Length;
        var headerTag = "h" + count;
        var headerHtml = Wrap(markdown.Substring(count + 1), headerTag);
        return (isList ? "</ul>" + headerHtml : headerHtml, false);
    }

    private static (string result, bool isList) ParseLineItem(string markdown, bool isList)
    {
        var innerHtml = Wrap(ParseText(markdown.Substring(2), true), "li");
        return (isList ? innerHtml : "<ul>" + innerHtml, true);
    }

    private static (string result, bool isList) ParseParagraph(string markdown, bool isList)
    {
        var parsed = ParseText(markdown, isList);
        return (isList ? "</ul>" + parsed : parsed, false);
    }

    private static (string result, bool isList) ParseLine(string markdown, bool isList)
    {
        if(markdown.StartsWith("#"))
            return ParseHeader(markdown, isList);
        else if(markdown.StartsWith("*"))
            return ParseLineItem(markdown, isList);
        else 
            return ParseParagraph(markdown, isList);
    }

    public static string Parse(string markdown)
    {
        var (result, isList) = markdown.Split('\n')
            .Aggregate(("", false), (o, line) => 
            {
                var (current, stillList) = o;
                var (lineResult, listResult) = ParseLine(line, stillList);
                return (current + lineResult, listResult);
            });

        return isList ? result + "</ul>" : result;
    }
}

File: fsharp/markdown/Markdown.fs
module Markdown

let parse (markdown:string) =
    let rec parse (text:string) =
        if text.Length < 2 then text else 
        match Seq.take 2 text |> Seq.toList with
        | ['_';'_'] -> 
            let infix = text.[2..].IndexOf("__") + 1
            let target = text.[2..infix]
            let next = text.[target.Length + 4..]
            sprintf "<strong>%s</strong>%s" (parse target) (parse next)
        | ['_';_] -> 
            let infix = text.[1..].IndexOf("_")
            let target = text.[1..infix]
            let next = text.[target.Length + 2..]
            sprintf "<em>%s</em>%s" (parse target) (parse next)
        | ['*';' '] -> sprintf "<li>%s</li>" (parse text.[2..])
        | ['#';_] ->
            let size = Seq.filter ((=) '#') text.[0..5] |> Seq.length
            sprintf "<h%i>%s</h%i>" size (parse text.[size + 1..]) size
        | _ -> 
            let infix = text.IndexOf('_')
            if infix < 1 then text else text.[0..infix - 1] + (parse text.[infix..])

    let folder (html,isList) line = 
        let parsed = parse line
        match Seq.take 2 parsed |> Seq.toList, isList with
        | ['<'; 'l' ], true -> (html + parsed, true)
        | ['<'; 'l' ], false -> (html + "<ul>" + parsed, true)
        | ['<'; 'h' ], true -> (html + "</ul>" + parsed, false)
        | ['<'; 'h' ], false -> (html + parsed, false)
        | [_;_], true -> (html + "</ul><p>" + parsed + "</p>", false)
        | _ -> (html + "<p>" + parsed + "</p>", false)

    let state = markdown.Split('\n') |> Seq.fold folder ("", false)
    if snd state then fst state + "</ul>" else fst state